### PR TITLE
Add Turbo stream morph action

### PR DESCRIPTION
**Pull request by omarluq**
This PR introduces a morph Turbo Stream action. The motivation for this addition stems from the recent discussions and suggestions in issue [#1163](https://github.com/hotwired/turbo/issues/1163). There is a clear interest in providing users with more flexible options for DOM manipulation, especially for those not ready or looking to fully transition to using Turbo's full page morphing capabilities.

turbo-rails PR https://github.com/hotwired/turbo-rails/pull/583
documentation PR https://github.com/hotwired/turbo-site/pull/169

### CODE REVIEW

File: src/core/drive/morph_renderer.js [hunk 1]
```diff
@@ -57,9 +57,9 @@ export class MorphRenderer extends PageRenderer {
         })
 
         return !event.defaultPrevented
-      } else {
-        return false
       }
+      
+      return false
```
Comment (omarluq) [inline src/core/drive/morph_renderer.js:62]: The `else` was redundant here 🤧

Comment (jorgemanrubia) [reply src/core/drive/morph_renderer.js:62]: Hey @omarluq, you are right it's redundant but I'd rather keep the PR focused on its purpose without performing cosmetic changes around 🙏.

**Review by jorgemanrubia — CHANGES_REQUESTED** [review]
Hey @omarluq, thanks for the PR:

I think a morph stream action makes sense but I am not sure about those `data-` attributes. It's quite a new API the one we would be introducing. I'd aim to start with something simple, like what [`turbo-morph`](https://github.com/marcoroth/turbo-morph) does, where it only supports a "children only" option via a regular attribute in the stream action, which matches morphdom. I'd start without any option here. 

This would also need to be consistent with the [events](https://turbo.hotwired.dev/reference/events#page-refreshes) we dispatch during morphing, both in the events it emits and in how they let you cancel the operation. We should probably extract the morphing operation out to a common place we can reuse.

Finally, this would need its counterpart helper in [`turbo-rails`](https://github.com/hotwired/turbo-rails) and documentation in [`turbo-site`](https://github.com/hotwired/turbo-site).

**Comment by omarluq** [comment]
Hey @jorgemanrubia, thank you for the speedy review sir! I've dropped the cosmetic change 🫡. 
Extracting the morph operation into a common place sound like a great idea; how about helper function in `util.js`? open to other suggestions as well. 
Also, I opened a related PR in turbo-rails, but there are a few specs that aren't passing. I plan to tackle those tomorrow along with updates you suggested to simplify the API! Ty much 🙏🏻

**Comment by jorgemanrubia** [comment]
> Extracting the morph operation into a common place sound like a great idea; how about a helper function in util.js? open to other suggestions as well.

Checking the code, this might not be the best path, because the current renderer has some parts that are tied to a Turbo page rendering operation. I'd see how the code looks without doing this first.

File: src/core/streams/stream_actions.js [hunk 2]
```diff
@@ -36,5 +38,87 @@ export const StreamActions = {
...
+          newElement: this.templateContent,
+        })
+      } catch (error) {
+        console.error(error)
+      }
+    })
+  },
+}
+
+const dispatchEvent = (eventName, detail, cancelable = false) => {
```
Comment (marcelokanzaki) [inline src/core/streams/stream_actions.js:69]: `dispatchEvent` is a js API https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent

Also, you might want to follow their conventions and use `dispatch` directly ☺️

Comment (omarluq) [reply src/core/streams/stream_actions.js:69]: Hey @marcelokanzaki, I am aware of that, but this is scoped locally, so it won't collide with the global `dispatchEvent`. The file is small and clean, and I'm trying to keep it as tight as possible. I didn't want to repeat the same object all over the place; it just doesn't look as good 😄 however if there is a strong objection against this pattern Im open to changing it.

Comment (omarluq) [reply src/core/streams/stream_actions.js]: @marcelokanzaki I took some time to consider your feedback, ended up refactoring to your suggestion 👍🏻

**Comment by omarluq** [comment]
Hey @jorgemanrubia, I believe I've addressed all of your feedback. Additionally, I added tests for the data-turbo-morph-style="innerHTML" option. I would appreciate a review. 🫡

File: src/tests/unit/stream_element_tests.js [hunk 3]
```diff
@@ -196,3 +196,33 @@ test("test action=refresh discarded when matching request id", async () => {
...
+  target.setAttribute("data-turbo-morph-style", "innerHTML")
+
+  subject.append(element)
+
+  await nextAnimationFrame()
+
+  assert.ok(subject.find("div#hello"))
+  assert.ok(subject.find("div#hello > h1#hello-child-element"))
+  assert.equal(subject.find("div#hello > h1#hello-child-element").textContent, "Hello Turbo Morphed")
+})
```
Comment (jorgemanrubia) [inline src/tests/unit/stream_element_tests.js:227]: We'd also need tests covering how Stimulus listeners can handle events. See tests in `page_refresh_tests.js`.

Comment (omarluq) [reply src/tests/unit/stream_element_tests.js:227]: Hey @jorgemanrubia, could you expand a little on what exactly we are looking to test, targeting a controller element with a morph stream action? or just a controller listening to the morph events?

Comment (omarluq) [reply src/tests/unit/stream_element_tests.js:227]: Hey @jorgemanrubia, just checking in to see if you've had a chance to look at my previous question. It's the final piece I need to wrap up this PR. Thanks a lot!

Comment (jorgemanrubia) [reply src/tests/unit/stream_element_tests.js:227]: Sorry about the delay here @omarluq. What we need to test are two things:

- That the stream action triggers the events we expect. You can see an example of how we do that in the general case here:

https://github.com/hotwired/turbo/blob/e8d25b29460d745955dc9c0f11f046eac54996dc/src/tests/functional/page_refresh_tests.js#L38-L45

- That the `before-` events that the stream action emits give the listener a chance to cancel the action. Example here:

https://github.com/hotwired/turbo/blob/e8d25b29460d745955dc9c0f11f046eac54996dc/src/tests/functional/page_refresh_tests.js#L47-L60

The PR is looking great. Thanks for your work here.

Comment (omarluq) [reply src/tests/unit/stream_element_tests.js:227]: @jorgemanrubia No worries, thank you for your guidance, I pushed new browser test to cover both. Please lmk if you have anymore feedback.

File: src/core/streams/actions/morph.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        morphStyle: morphStyle,
+        callbacks: {
+          beforeNodeAdded,
+          beforeNodeMorphed,
+          beforeAttributeUpdated,
+          beforeNodeRemoved,
+          afterNodeMorphed
+        }
+      })
+    } catch (e) {
```
Comment (jorgemanrubia) [inline src/core/streams/actions/morph.js:18]: Omar could we remove the wrapping `try/catch`, please? I don't think this should raise any exception so we would be silencing bugs we don't know about.

**Review by jorgemanrubia — COMMENTED** [review]
Thanks @omarluq. Another little comment about the removing the wrapping try/catch and we'll get this merged.

cc @afcapel for another pair of eyes here.

File: src/core/streams/actions/morph.js (hunk 4 above)
Comment (omarluq) [reply src/core/streams/actions/morph.js]: @jorgemanrubia removed

File: src/tests/functional/morph_stream_action_tests.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  await nextEventOnTarget(page, "message_1", "turbo:before-morph-element")
+  await nextEventOnTarget(page, "message_1", "turbo:morph-element")
+  await expect(page.locator("#message_1")).toHaveText("Morphed")
+})
+
+test("preventing a turbo:before-morph-element prevents the morph", async ({ page }) => {
+  await page.goto("/src/tests/fixtures/morph_stream_action.html")
+
+
```
Comment (afcapel) [inline src/tests/functional/morph_stream_action_tests.js:27]: Please remove the extra CR here.

**Review by afcapel — COMMENTED** [review]
This looks good to me. Just a minor formatting issue.

File: src/tests/functional/morph_stream_action_tests.js (hunk 5 above)
Comment (omarluq) [reply src/tests/functional/morph_stream_action_tests.js]: @afcapel good eye for details 🔥 removed 🫡

File: src/core/streams/actions/morph.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+function afterNodeMorphed(target, newElement) {
+  if (newElement instanceof HTMLElement) {
+    dispatch("turbo:morph-element", {
+      target,
+      detail: {
+        newElement
+      }
+    })
+  }
```
Comment (seanpdoyle) [inline src/core/streams/actions/morph.js:62]: @jorgemanrubia duplicating the `MorphRenderer` logic feels like a big risk. Would generalizing the details of how we invoke Idiomorph (through something like https://github.com/hotwired/turbo/pull/1192) help here?

Comment (omarluq) [reply src/core/streams/actions/morph.js:62]: I tend to agree with @seanpdoyle I think [`morph_elements.js`](https://github.com/hotwired/turbo/blob/07c35779f63b4261f37542eaba6ff62fe741c927/src/core/morph_elements.js) can be a drop in replacement for this class

Comment (jorgemanrubia) [reply src/core/streams/actions/morph.js:62]: @seanpdoyle the problem is that the current implementation of the page renderer has logic interweaved to deal with morphing frames. I think we need to untangle that in https://github.com/hotwired/turbo/pull/1192, but I still don't see a clear separation there, and I haven't had proper time for looking into that yet. Agree that, ultimately, we should invoke the same morphing logic from everywhere, but we need to do some deep refactor to accommodate that.

Comment (seanpdoyle) [reply src/core/streams/actions/morph.js:62]: > to deal with morphing frames

How does the exclusion of the Frame handling logic from the MorphRenderer impact the Stream's behavior?

Are there downsides to handling Frames consistently across Drive navigation morphs and Stream morphs?

Comment (seanpdoyle) [reply src/core/streams/actions/morph.js:62]: @jorgemanrubia In response to https://github.com/hotwired/turbo/pull/1185#discussion_r1527139119, I've opened https://github.com/hotwired/turbo/pull/1234 to share the bulk of the morphing logic across Pages, Frames, and Streams.

**Pull request merged**